
# Experiment 2: Data Pre-processing & EDA Lab  
## Use Case: Patient Health Monitoring System

Healthcare datasets often contain missing values, categorical variables, and outliers.  
This experiment focuses on preparing medical data for reliable analysis and machine learning.

### Dataset Used
**Heart Disease Dataset**  
Source: Kaggle  
https://www.kaggle.com/datasets/johnsmith88/heart-disease-dataset

**Tasks:**
- Identify missing values
- Apply mean/median/mode imputation
- k-NN based imputation

Outcome: Reliable preprocessing of medical datasets

##1. Tools and Libraries Used

In [ ]:
# Import required libraries

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.impute import SimpleImputer, KNNImputer


## 2. Data Acquisition – Loading the Dataset


In [ ]:
# Load the dataset

df = pd.read_csv("/content/heart.csv")
df.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,52,1,0,125,212,0,1,168,0,1.0,2,2,3,0
1,53,1,0,140,203,1,0,155,1,3.1,0,0,3,0
2,70,1,0,145,174,0,1,125,1,2.6,0,0,3,0
3,61,1,0,148,203,0,1,161,0,0.0,2,1,3,0
4,62,0,0,138,294,1,1,106,0,1.9,1,3,2,0



## 3. Understanding the Dataset


In [ ]:
df.shape

(1025, 14)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1025 entries, 0 to 1024
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1025 non-null   int64  
 1   sex       1025 non-null   int64  
 2   cp        1025 non-null   int64  
 3   trestbps  1025 non-null   int64  
 4   chol      1025 non-null   int64  
 5   fbs       1025 non-null   int64  
 6   restecg   1025 non-null   int64  
 7   thalach   1025 non-null   int64  
 8   exang     1025 non-null   int64  
 9   oldpeak   1025 non-null   float64
 10  slope     1025 non-null   int64  
 11  ca        1025 non-null   int64  
 12  thal      1025 non-null   int64  
 13  target    1025 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 112.2 KB


**Observation**

Dataset contains 1025 records and 14 attributes

No explicit NaN values are visible

In [ ]:
df.describe()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
count,1025.000000,1025.000000,1025.000000,1025.000000,1025.00000,1025.000000,1025.000000,1025.000000,1025.000000,1025.000000,1025.000000,1025.000000,1025.000000,1025.000000
mean,54.434146,0.695610,0.942439,131.611707,246.00000,0.149268,0.529756,149.114146,0.336585,1.071512,1.385366,0.754146,2.323902,0.513171
std,9.072290,0.460373,1.029641,17.516718,51.59251,0.356527,0.527878,23.005724,0.472772,1.175053,0.617755,1.030798,0.620660,0.500070
min,29.000000,0.000000,0.000000,94.000000,126.00000,0.000000,0.000000,71.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,48.000000,0.000000,0.000000,120.000000,211.00000,0.000000,0.000000,132.000000,0.000000,0.000000,1.000000,0.000000,2.000000,0.000000
50%,56.000000,1.000000,1.000000,130.000000,240.00000,0.000000,1.000000,152.000000,0.000000,0.800000,1.000000,0.000000,2.000000,1.000000
75%,61.000000,1.000000,2.000000,140.000000,275.00000,0.000000,1.000000,166.000000,1.000000,1.800000,2.000000,1.000000,3.000000,1.000000
max,77.000000,1.000000,3.000000,200.000000,564.00000,1.000000,2.000000,202.000000,1.000000,6.200000,2.000000,4.000000,3.000000,1.000000



## 4. Identifying Missing Values

Why missing values occur?

- Sensor failure
- Human error
- Incomplete medical records


In [ ]:
df.isnull().sum()

,0
age,0
sex,0
cp,0
trestbps,0
chol,0
fbs,0
restecg,0
thalach,0
exang,0
oldpeak,0


All columns show **0 missing values.**

**Why did this happen?**

This happens because many real-world datasets (especially medical datasets) do not store missing data as NaN.

Instead, missing or unknown values are often represented as:

- 0

- ?

- -1

- Empty strings ('')

**isnull() only detects NaN / None, not logically missing values.**

**Columns that MAY contain logical missing values** (only if 0 exists in data)

- trestbps (Resting blood pressure)

- chol (Serum cholesterol)

- thalach (Maximum heart rate)

In [ ]:
#Detect Logical Missing Values

#Example: Cholesterol cannot be zero

(df['chol'] == 0).sum()

np.int64(0)

In [ ]:
#Example: Blood pressure cannot be zero

(df['trestbps'] == 0).sum()

np.int64(0)

In [ ]:
#Example: thalach (Maximum heart rate)

(df['thalach'] == 0).sum()

np.int64(0)

In [ ]:
df[['chol', 'trestbps', 'thalach']].nunique()

,0
chol,152
trestbps,49
thalach,91


In [ ]:
df['chol'].unique()

array([212, 203, 174, 294, 248, 318, 289, 249, 286, 149, 341, 210, 298,
       204, 308, 266, 244, 211, 185, 223, 208, 252, 209, 307, 233, 319,
       256, 327, 169, 131, 269, 196, 231, 213, 271, 263, 229, 360, 258,
       330, 342, 226, 228, 278, 230, 283, 241, 175, 188, 217, 193, 245,
       232, 299, 288, 197, 315, 215, 164, 326, 207, 177, 257, 255, 187,
       201, 220, 268, 267, 236, 303, 282, 126, 309, 186, 275, 281, 206,
       335, 218, 254, 295, 417, 260, 240, 302, 192, 225, 325, 235, 274,
       234, 182, 167, 172, 321, 300, 199, 564, 157, 304, 222, 184, 354,
       160, 247, 239, 246, 409, 293, 180, 250, 221, 200, 227, 243, 311,
       261, 242, 205, 306, 219, 353, 198, 394, 183, 237, 224, 265, 313,
       340, 259, 270, 216, 264, 276, 322, 214, 273, 253, 176, 284, 305,
       168, 407, 290, 277, 262, 195, 166, 178, 141])

In [ ]:
df[['chol', 'trestbps', 'thalach']] = df[['chol', 'trestbps', 'thalach']].replace(0, np.nan)

In [ ]:
df.isnull().sum()

,0
age,0
sex,0
cp,0
trestbps,0
chol,0
fbs,0
restecg,0
thalach,0
exang,0
oldpeak,0


**Interpretation:**


The count is 0 → no logical missing values present in this dataset


## 5. Statistical Imputation of Missing Values


**1. What is Imputation?**

Imputation is the process of replacing missing values with meaningful estimates so that:

- The dataset becomes complete

- Statistical analysis and ML algorithms can be applied

- Information loss is minimized

**2. Why Imputation is Required**

Most machine learning algorithms:

- Cannot handle missing values

- Produce errors or biased results if NaN values are present

Therefore, missing values must be handled before further analysis.

**3. Types of Statistical Imputation**

Statistical imputation replaces missing values using summary statistics of the data:

- Mean(Average): Numerical (normally distributed)

- Median(Middle value):	Numerical (skewed / outliers)

- Mode(Most frequent value):	Categorical / discrete

To demonstrate imputation techniques, we introduce missing values artificially (for learning purposes).

In [ ]:
# Introduce missing values artificially

df_missing = df.copy()
df_missing.loc[5:15, 'chol'] = np.nan
df_missing.loc[0:8, 'trestbps'] = np.nan

df_missing.isnull().sum()

,0
age,0
sex,0
cp,0
trestbps,9
chol,11
fbs,0
restecg,0
thalach,0
exang,0
oldpeak,0


In [ ]:
# Mean Imputation

mean_imputer  = SimpleImputer(strategy='mean')
df_missing['chol'] = mean_imputer.fit_transform(df_missing[['chol']])

'''OR
df_missing['chol'].fillna(df_missing['chol'].mean(), inplace=True)'''

"OR\ndf_missing['chol'].fillna(df_missing['chol'].mean(), inplace=True)"

In [ ]:
df_missing.isnull().sum()

,0
age,0
sex,0
cp,0
trestbps,9
chol,0
fbs,0
restecg,0
thalach,0
exang,0
oldpeak,0


In [ ]:
# Median Imputation

median_imputer = SimpleImputer(strategy='median')
df_missing['trestbps'] = median_imputer.fit_transform(df_missing[['trestbps']])

'''OR
df_missing['trestbps'].fillna(df_missing['trestbps'].median(), inplace=True)'''


"OR\ndf_missing['trestbps'].fillna(df_missing['trestbps'].median(), inplace=True)"

In [ ]:
#Mode Imputation (Example)

mode_imputer = SimpleImputer(strategy='most_frequent')
df_missing['Gender'] = mode_imputer.fit_transform(df_missing[['Gender']])

'''OR
df_missing['Gender'].fillna(df_missing['Gender'].mode()[0], inplace=True)'''

"OR\ndf_missing['sex'].fillna(df_missing['sex'].mode()[0], inplace=True)"


## 6. k-NN Based Data Imputation


**1. What is k-NN Imputation?**

k-Nearest Neighbors (k-NN) Imputation is an advanced missing-value handling technique that:

- Replaces missing values using similar data points

- Uses the idea that similar patients have similar medical measurements

Unlike mean/median imputation, k-NN preserves relationships between features.

**For a patient record with a missing value:**

- Find k most similar patients (neighbors)

- Measure similarity using distance metrics

- Replace the missing value with:

Mean of neighbors (numerical)

Most frequent value (categorical)

In [ ]:
# Introduce missing values artificially

df_knn = df.copy()
df_knn.loc[5:15, 'chol'] = np.nan
df_knn.loc[0:8, 'trestbps'] = np.nan

df_knn.isnull().sum()

,0
age,0
sex,0
cp,0
trestbps,9
chol,11
fbs,0
restecg,0
thalach,0
exang,0
oldpeak,0


In [ ]:
#Rows with missing values

df_knn[df_knn.isnull().any(axis=1)]

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,52,1,0,NaN,212.0,0,1,168,0,1.0,2,2,3,0
1,53,1,0,NaN,203.0,1,0,155,1,3.1,0,0,3,0
2,70,1,0,NaN,174.0,0,1,125,1,2.6,0,0,3,0
3,61,1,0,NaN,203.0,0,1,161,0,0.0,2,1,3,0
4,62,0,0,NaN,294.0,1,1,106,0,1.9,1,3,2,0
5,58,0,0,NaN,NaN,0,0,122,0,1.0,1,0,2,1
6,58,1,0,NaN,NaN,0,2,140,0,4.4,0,3,1,0
7,55,1,0,NaN,NaN,0,0,145,1,0.8,1,1,3,0
8,46,1,0,NaN,NaN,0,0,144,0,0.8,2,0,3,0
9,54,1,0,122.0,NaN,0,0,116,1,3.2,1,2,2,0


In [ ]:
knn_imputer = KNNImputer(n_neighbors=5)    #The missing value is filled using 5 most similar records

df_knn = pd.DataFrame(knn_imputer.fit_transform(df), columns=df.columns)
df_knn.head(15)

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,52.0,1.0,0.0,125.0,212.0,0.0,1.0,168.0,0.0,1.0,2.0,2.0,3.0,0.0
1,53.0,1.0,0.0,140.0,203.0,1.0,0.0,155.0,1.0,3.1,0.0,0.0,3.0,0.0
2,70.0,1.0,0.0,145.0,174.0,0.0,1.0,125.0,1.0,2.6,0.0,0.0,3.0,0.0
3,61.0,1.0,0.0,148.0,203.0,0.0,1.0,161.0,0.0,0.0,2.0,1.0,3.0,0.0
4,62.0,0.0,0.0,138.0,294.0,1.0,1.0,106.0,0.0,1.9,1.0,3.0,2.0,0.0
5,58.0,0.0,0.0,100.0,248.0,0.0,0.0,122.0,0.0,1.0,1.0,0.0,2.0,1.0
6,58.0,1.0,0.0,114.0,318.0,0.0,2.0,140.0,0.0,4.4,0.0,3.0,1.0,0.0
7,55.0,1.0,0.0,160.0,289.0,0.0,0.0,145.0,1.0,0.8,1.0,1.0,3.0,0.0
8,46.0,1.0,0.0,120.0,249.0,0.0,0.0,144.0,0.0,0.8,2.0,0.0,3.0,0.0
9,54.0,1.0,0.0,122.0,286.0,0.0,0.0,116.0,1.0,3.2,1.0,2.0,2.0,0.0


In [ ]:
df_knn.isnull().sum()

,0
age,0
sex,0
cp,0
trestbps,0
chol,0
fbs,0
restecg,0
thalach,0
exang,0
oldpeak,0
